### DISCUSSION TITLE
How to disable collections for S3?

### DISCUSSION THREAD

**Discussion by ladserg** [seaweedfs/seaweedfs#5064] [Q&A]
When we create a bucket in S3 storage, a separate collection is created that requires separate volumes. Due to intensive work with the file system, all volumes are already allocated and an error occurs due to the lack of free volumes.

It is assumed that a large number of baskets (dozens) will be periodically created and deleted. The need to allocate separate volumes for each bucket creates a problem for us (there are often no free volumes). At the same time, it is difficult to plan the required disk space, since it is logically divided into volumes of different sizes.

Creating buckets in the same collection as the file system would solve the problem for us.

Is this possible?

**Comment by chrislusf** [comment]
can you reduce the volume size limit, to have more volumes?

**Reply by ladserg** [reply]
Yes. But unfortunately this does not solve the problem. I'm afraid that too small a volume size will harm the performance of the file system, both daily work files and backups are stored there. And the question remains by how much to reduce the size? We have baskets of 10 MB, and there are baskets of tens of GB. This heterogeneity creates a problem when there are a large number of baskets.

At the file system level, SeaweedFS solved for us the problem of combining disk capacity into a single space. Flexible scaling and distribution allows us to solve the problem of disks of different sizes on different servers.

We are currently looking at using S3 in SeaweedFS and dividing the buckets into collections (individual volumes) created a problem for us and brought us back to the problem of different sized disks.

**Reply by chrislusf** [reply]
Maybe later we can have small buckets to "attach" to a parent bucket, so that the small buckets data go to the parent collection.

Any other ideas?

**Comment by itsfarbod** [comment]
Actually I found a solution for this.
With another S3 instance which is configured to another path than the default one which is `/buckets` and some little changes in code we can set all buckets in that path to one collection(With the help of `fs.configure`). This issue happened to me when I wanted small size buckets. Even by setting the volume growth value to the smallest possible value(According to documentation), my default volume size in the ecosystem is 10Gb, so it means a bucket will take 10Gb of storage in it's initial state. So I think of a model which buckets are shared across their collection.
Some pros about this solution:
- It's my guess, please correct it if I'm wrong: If we set the initial value of volumes to something like 1Gb, in hi-cap systems(e.g like a seaweed with 1000Tb of storage) we will encounter performance issues. But with this change we can have buckets with custom sizes like 100Mb without decreasing the performance. The initial value of volumes can be 10 or even 100Gb and also we can have shared buckets with much lower capacities.
-  Even with limiting the volume growth value for a collection, initial capacity is still high for a system with default 10Gb size for volumes. But now we can have replication for even small buckets.

And here are cons about this change which came to my mind:
- It seems nature of the collections for buckets  is for deleting them much more quickly. Because you have to only delete some file in your volume servers(.dat .idx etc)and there is no recursive deletion which takes some time. But in this model we have to delete a bucket recursively in the volume servers. And the request to the filer won't be a deleteCollection, it will be a DeleteEntry  request with recursive enabled. 
- we have to make custom metrics for them. Like size etc.
- Separate S3 instance. 
- Expansion. Imagine the smallest size for volumes is 10Gb. What if the user create a small size bucket and then expand it to over 10Gb? In this case the initial bucket is in the shared collection but now it have to transform to standard buckets(With separate collection).

Here are all my solutions for small buckets:
- The solution that I provided above. Shared collections.
- Have another seaweed instance with all its component with lower volume size. Notice that this solution will have the volume count problem too and also the expansion problem.

And I have some questions:
- For volume expansion we can configure the paths to be able to change the collection. e.g first it will have a shared collection, then it will be configured to have separate collection. What problems we will encounter with this solution?
-  The latest version of code write now, isn't configurable for default buckets path(`/buckets` is default). How about to make it configurable? 
- Some parts of the code are hard coded for `/buckets`. e.g when you want to set a collection for a bucket path because it starts with `/buckets` you will encounter an error. But this error doesn't show when you set a custom path for buckets. How about to change this behavior?
- Why we can't have asymmetric volume sizes? I mean volumes with configurable size?

What have i done till now:
- There is no problem  with operations on shared collections in a path other than '/buckets' except deleting the bucket. So i changed the behavior of the S3 to have a flag which tells to create buckets in a different path and also make it shared and also implemented the deletion logic for recursively deleting the small size bucket in the collection.
- I tested it out, It doesn't seem to have problems.

And I can do a PR for current solution, I mean the shared collection approach.

**Comment by ladserg** [comment]
I apologize for my bad English. If I translated something incorrectly, please clarify.

**Advantage** SeaweedFS – volume servers. At any time, you can add new volume servers and remove unnecessary ones. SeweedFS Volume Server allows you to create small storage and scale it up to infinity in the future (theoretically). All of this is true as long as the volume servers are homogeneous.

It seems to me that many problems would not exist if volume servers were allocated to a separate abstraction layer, responsible only for storing blocks of data, no matter what content.

SeaweedFS Volume Server is very similar to OSD CEPH. OSD provides storage of data blocks and work with them. Other services are not interested in how the OSD stores these blocks. True, OSD does not provide data replication. Data replication is provided by a separate service at the pool level (in CEPH, replication is tied to the data pool).

This implementation greatly simplifies the management of physical storage devices.

Currently, SeaweedFS replicates data at the volume level, making it difficult to manage physical storage devices. The cleanup code requires free volumes to be available, and if they are not available, an error occurs and the volume that failed to replicate goes into the ReadOnly state, breaking the storage structure. Issues also arise when replicating volumes of different sizes.

This may not be the case for other communities or companies, but in all the cases I came across, they wanted to consolidate their disk resources located on their servers into a single managed space.

Most likely my case is not like everyone else's. My case is not aimed at the developer community, but at the corporate sector.

Let me give you an example of my problems. We have several dozen physical servers. These servers host:

* ProxMox cluster
* Kubernetes
* Various DBMS
* Backup server management data
* Various file and object storage
* Servers with mission-critical services that require high performance and availability.

All this requires storing data such as:
* User files
* Application and service data
* Backups
* Intermediate data of virtual machines and Kubernetes services of different categories.
* Other data

Now all this is divided into many different storages: NFS, SMB, Minio S3, etc.

At the file system level, SeaweedFS solves our problems until collections appear :-)

Unfortunately, I could not find information about what problems the collections solve. Volume servers become unstable when trying to create volumes of different types. The volume server does the right thing by creating new volumes for data in advance, this speeds up work at the data block level, but it breaks down if you need to divide the data at the volume server level.

I think the solution to the problem would be to change the concept of working with data. But I'm not sure that SeaweedFS can be brought to this point or that the developers will be interested in it.

The current structure of SeaweedFS is as follows:
* Master
* Filer (file and object management)
* Volume (management of data blocks, replications, collections).

If replications and collections are moved to a layer higher, to a separate logical level, this will strengthen the SeaweedFS structure. Volume servers will only be responsible for storing data at the block level.

Data replication must occur at a higher level, the replication layer must pass different copies of blocks to the volume servers. This separation will allow the replication modules and volumes to be maintained and developed separately.

It is better to divide the storage space at the logical level. At a level above replication. You can split it into collections and pools. It is better to bind replication methods not to files, but to collections (pools). This will greatly simplify the management of the shared space and make its operation more stable and reliable.

As a result, in the proposed structure, the file or object will go through the following path:
1. Master - he will tell you which Filer is free and available. He also knows what services are available. He doesn't know where or how the data is stored. He's not interested. This is not its function. Its task is to coordinate actions between individual services and modules. Perhaps it can be entrusted with centralized storage of metadata, following the example of Consul.
2. Filer – will provide the necessary data transfer protocol. Determines which collection the file or object is in. Controls access to files/objects and quotas. Manages the life cycle of files and objects. It knows which files and objects are stored in which collections. But it doesn't care about data synchronization and storage. Others do it. It only cares about data management at the file and object level. It also encrypts the data and transmits it encrypted to the Pooler service.
3. Pooler – controls the division of space into collections (pools), defines such parameters as the replication method, restrictions on the size of the collection. It doesn't know anything about files, objects, quotas. He doesn't know how his data is stored. It accepts blocks of data from Filer. Depending on the replication settings, Pooler transfers the required number of copies of data blocks to volume servers.
4. Volume server - knows nothing about files, objects, replication, collections (pools). It only stores blocks of data. They give him a block of data - he somehow saves it, they ask him for a block of data - he gives it back. Blocks of data can be stored in different ways. Data blocks can be stored in separate volumes of different sizes using a block device (HDD, SDD, LVM, iSCSI). A volume server can even store its data in object storage (S3, CEPH, etc.).

This structure, in my humble opinion, will increase the reliability, manageability and stability of the storage. If anyone is interested :-)

**Reply by chrislusf** [reply]
How about optionally removing the one-bucket-one-collection requirement?

The collection deletion would be slower.

**Reply by ladserg** [reply]
Interesting. How does an S3 bucket differ from a large directory located on the file system? How is this solved when deleting a large directory located on the file system?

Perhaps I have incomplete information about the collections. So far they are puzzling me. Disabling collections could be a temporary solution.

**Reply by itsfarbod** [reply]
As I figured out, A collection is a set of volumes. Volumes in seaweed are just some files(.dat .idx) on the volume server. When you deleting a collection(In default implementation, a bucket has a collection) all its volumes which are some files on the volume server will be deleted. But when you want to delete a file from a collection, It have to find it in the volume files and delete it which is slower than deleting a volume(I think it's because of indexing and other things)
So the concept of collections is  kinda interesting.
But you can't have volumes with different sizes. So All buckets which actually are collections and collection actually are just some  files, will have the same size which it's initial size can be calculated using volume size and volume growth values. 
As Lu said, we can delete one-bucket-one-collection approach, which means we have to put some buckets in only one collection.
But as I mentioned above, the deletion will be slower I think. But in this case you have buckets with arbitrary sizes.

**Reply by ladserg** [reply]
Generally, deleting large amounts of data is rare. At least this is true for distributed storage.

If the volume, along with the data block, stores information about the metadata of the files, then this explains the slow deletion. If metadata about files is stored by a filer, then the operation of deleting any number of objects is equivalent to the operation of deleting files.

Alternatively, data blocks can be marked as free and released in a vacuum.

**Reply by INeedANameToo** [reply]
> How about optionally removing the one-bucket-one-collection requirement?

@chrislusf Sorry to bump on such an old thread, but was this ever expanded upon? Is there some flag or configuration supported today that lets us disable this one-bucket-per-collection behavior?

**Reply by chrislusf** [reply]
Still one-bucket-per-collection. What is your use case for changing this?

**Reply by ladserg** [reply]
We started using a different program. The problem is no longer relevant to us.

**Reply by INeedANameToo** [reply]
> Still one-bucket-per-collection. What is your use case for changing this?

I was hoping to use seaweed with a flexible bucket structure. Specifically, my use case potentially has many buckets each with varying space utilization (mix of large buckets and small buckets). 

IIUC, if i run with a high-limit volume count, I lose a bit of functionality balancing across volume servers. 

And if I run with a constrained volume count spread across multiple collections, I risk low-usage buckets leaving under utilized volumes.

So I was hoping to be able to map all buckets to a shared collection. 

Besides the bucket deletion speed tradeoff and the loss of collection-level tenancy, (which are not that relevant to my use case) is there any obvious downsides to doing this?

**Reply by chrislusf** [reply]
you can use a low volume size limit such as 1GB.

**Reply by INeedANameToo** [reply]
In my use case where I am creating lots of tiny buckets and a few big ones, the tiny buckets can still be space inefficient. 

And IIUC, creating volume structures are not free + a lot of seaweed operations scale with the number of volumes. 

if I make the volume size too small, I could end up with too many volumes and make my large buckets very inefficient.
